### DISCUSSION TITLE
Trouble connecting to MediaMTX WHEP server with Python client

### DISCUSSION THREAD

**Discussion by szym1998** [bluenviron/mediamtx#3640] [Questions]
### Question

I'm trying to create a Python client to connect to a MediaMTX server using the WHEP (WebRTC-HTTP Egress Protocol) protocol, but I'm struggling to get started. I have a MediaMTX server running, but I'm not sure how to properly implement the client-side WHEP connection in Python.

My current setup:
- MediaMTX server running on http://10.130.7.166:8889/right/whep (the webpage with stream opens with http://10.130.7.166:8889/right)
- Python environment with aiortc and aiohttp libraries installed

I've tried implementing a basic WHEP client based on the protocol description, but I'm not sure if I'm doing it correctly.

Can someone provide guidance on:
1. The correct sequence of HTTP requests for WHEP (POST for offer, PATCH for answer)?
2. How to properly handle the SDP exchange in Python?
3. Any example code or resources for a working WHEP client in Python?

Any help or pointers in the right direction would be greatly appreciated. Thank you!

**Comment by iconoclasthero** [comment]
I'm trying to figure out how to do WHEP to stream audio in opus (=no flv=no rtmp).  How did you create the whep stream?
(I want to catch it with ffmpeg or another existing cli tool.)

**Comment by aler9** [comment]
I did some experiments with aiortc and it's perfectly possible to receive a WebRTC+WHEP stream by converting an existing WHEP client, like [the one embedded inside MediaMTX](https://github.com/bluenviron/mediamtx/blob/main/internal/servers/webrtc/reader.js), from JavaScript to Python. Working example:


```py
import re
import asyncio
import aiohttp
from aiortc import RTCPeerConnection, RTCConfiguration, RTCIceServer, RTCSessionDescription

WHEP_URL = 'http://localhost:8889/stream/whep'

async def run_track(track):
    while True:
        frame = await track.recv()
        print(frame)

async def main():
    async with aiohttp.ClientSession() as session:
        # get ICE servers
        iceServers = []
        async with session.options(WHEP_URL) as res:
            for k, v in res.headers.items():
                if k == "Link":
                    m = re.match('^<(.+?)>; rel="ice-server"(; username="(.*?)"; credential="(.*?)"; credential-type="password")?', v)
                    iceServers.append(RTCIceServer(urls=m[1], username=m[2], credential=m[3], credentialType='password'))

        # setup peer connection
        pc = RTCPeerConnection(RTCConfiguration(iceServers=iceServers))
        pc.addTransceiver('video', direction='recvonly')
        pc.addTransceiver('audio', direction='recvonly')

        # on connection state change callback
        @pc.on("connectionstatechange")
        async def on_connectionstatechange():
            print("Connection state is %s" % pc.connectionState)

        # on track callback
        @pc.on('track')
        async def on_track(track):
            print('on track')
            task = asyncio.ensure_future(run_track(track))

        # generate offer
        offer = await pc.createOffer()
        await pc.setLocalDescription(offer)

        # send offer, set answer
        async with session.post(WHEP_URL, headers={'Content-Type': 'application/sdp'}, data=offer.sdp) as res:
            answer = await res.text()
        await pc.setRemoteDescription(RTCSessionDescription(sdp=answer, type='answer'))

        await asyncio.sleep(10000)

asyncio.run(main())
```

What is missing is the ability to send asynchronous candidates to the server (i.e. trickle ICE) - if you need that, you can go in in the conversion task from JavaScript to Python.
